source/server/hot_restarting_child.h

```diff
@@ -18,7 +18,8 @@ class HotRestartingChild : HotRestartingBase {
   // to support QUIC packet forwarding.
   class UdpForwardingContext {
   public:
-    using ForwardEntry = std::pair<const Network::Address::Instance*, Network::UdpListenerConfig*>;
+    using ForwardEntry = std::pair<Network::Address::InstanceConstSharedPtr,
```

Comment:
I would prefer not to at this time since `listener_worker_routers_` could also have benefited from doing that, and didn't do it - I'd tend to assume there's a reason for that.

It does look like Address::Instance has an equality operator that would mostly do the right thing, but given earlier conversations about dynamic_cast on the previous PR, there might be some issue around dynamic_cast e.g. [here](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/network/address_impl.cc?L310-313) where if true dynamic_cast isn't available then this would provoke an error if you were to look up an ipv4 address in a hash_map where it shares a hash with an ipv6 addresses.

Maybe worth filing an issue to update both things to key on an InstanceConstSharedPtr though.

Reply:
> I would prefer not to at this time since `listener_worker_routers_` could also have benefited from doing that, and didn't do it - I'd tend to assume there's a reason for that.
> 
> It does look like Address::Instance has an equality operator that would mostly do the right thing, but given earlier conversations about dynamic_cast on the previous PR, there might be some issue around dynamic_cast e.g. [here](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/network/address_impl.cc?L310-313) where if true dynamic_cast isn't available then this would provoke an error if you were to look up an ipv4 address in a hash_map where it shares a hash with an ipv6 addresses.
> 

This depends on how you implement Hash. But filing a issue for both and handle it when needed sounds good.

Reply:
Also would be awkward if you keyed on shared pointers, when you want to find() an address and what you have is a reference or local instance. I don't *think* there's a way to use alternative hash and equality operators for different candidate types.

Reply:
#29529 for a tracking issue.